radixdlt-java/src/main/java/com/radixdlt/client/core/network/epics/FindANodeEpic.java

```diff
@@ -0,0 +1,103 @@
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Epic which finds a connected sharded node when a FindANode request is received. If there are none found,
+ * then the epic attempts to start connections.
+ */
+public class FindANodeEpic implements RadixNetworkEpic {
+
+	private static final Logger LOGGER = LoggerFactory.getLogger(FindANodeEpic.class);
+	private final RadixPeerSelector selector;
+
+	public FindANodeEpic(RadixPeerSelector selector) {
+		this.selector = selector;
+	}
+
+	private Maybe<RadixNodeAction> findConnection(Set<Long> shards, RadixNetworkState state) {
+		final Map<WebSocketStatus, List<RadixNode>> statusMap = Arrays.stream(WebSocketStatus.values())
+			.collect(Collectors.toMap(
+				Function.identity(),
+				s -> state.getNodes().entrySet().stream()
+					.filter(entry -> entry.getValue().getShards().map(sh -> sh.intersects(shards)).orElse(false))
+					.filter(e -> e.getValue().getStatus().equals(s))
+					.map(Entry::getKey)
+					.collect(Collectors.toList())
+			));
+
+		final long activeNodeCount = statusMap.get(WebSocketStatus.CONNECTED).size() + statusMap.get(WebSocketStatus.CONNECTING).size();
+
+		if (activeNodeCount < 1) {
+			LOGGER.info(String.format("Requesting more node connections, want %d but have %d active nodes", 1, activeNodeCount));
+
+			List<RadixNode> disconnectedPeers = statusMap.get(WebSocketStatus.DISCONNECTED);
+			if (disconnectedPeers.isEmpty()) {
+				LOGGER.info("Could not connect to new peer, don't have any.");
+			} else {
+				return Maybe.just(ConnectWebSocketAction.of(disconnectedPeers.get(0)));
+			}
+		}
+
+		return Maybe.empty();
+	}
+
+	private static List<RadixNode> getConnectedNodes(Set<Long> shards, RadixNetworkState state) {
+		return state.getNodes().entrySet().stream()
+			.filter(entry -> entry.getValue().getStatus().equals(WebSocketStatus.CONNECTED))
+			.filter(entry -> entry.getValue().getShards().map(s -> s.intersects(shards)).orElse(false))
+			.map(Map.Entry::getKey)
+			.collect(Collectors.toList());
+	}
+
+	@Override
+	public Observable<RadixNodeAction> epic(Observable<RadixNodeAction> actions, Observable<RadixNetworkState> stateObservable) {
+		return actions
+			.filter(a -> a instanceof FindANodeRequestAction)
+			.map(FindANodeRequestAction.class::cast)
+			.flatMap(a -> {
+				Observable<RadixNetworkState> syncNetState = stateObservable
+					.replay(1)
+					.autoConnect(2);
+
+				Observable<List<RadixNode>> connectedNodes = syncNetState
+					.map(state -> getConnectedNodes(a.getShards(), state))
+					.replay(1)
+					.autoConnect(2);
+
+				// Try and connect if there are no nodes
+				Observable<RadixNodeAction> newConnections = syncNetState
+					.zipWith(connectedNodes.takeWhile(List::isEmpty), (s, n) -> s)
+					.flatMapMaybe(state -> findConnection(a.getShards(), state));
+
+				Observable<RadixNodeAction> selectedNode = connectedNodes
+					.filter(viablePeerList -> !viablePeerList.isEmpty())
```

Comment:
Where did the filters go? I don't see them used anywhere, but they're still here. Previously they were applied here

Reply:
Yes, kept them for reference so that they can be implemented later. Removed their use as they seem to expose a limiting interface which doesn't allow a filter to make decisions based on time.

Reply:
Sounds reasonable, what are you thinking of that will integrate time as well?

Reply:
e.g. selecting a node which has been online for atleast X amount of time or has not dropped out in the past minute.